### ISSUE TITLE
Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory

### ISSUE THREAD

**Issue by domindominik** [tensorflow/tensorflow#38578]
Debian 10.3
Python 3.8.2
Cuda 10.2
Quadro K5200d drivers 440.82
Tensorflow 2.2.0-rc2

import tensorflow as tf
tf.config.list_physical_devices('GPU')
2020-04-15 20:58:07.177217: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2020-04-15 20:58:07.205025: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties: 
pciBusID: 0000:03:00.0 name: Quadro K5200 computeCapability: 3.5
coreClock: 0.771GHz coreCount: 12 deviceMemorySize: 7.94GiB deviceMemoryBandwidth: 179.05GiB/s
2020-04-15 20:58:07.205269: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory
2020-04-15 20:58:07.207704: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2020-04-15 20:58:07.210137: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2020-04-15 20:58:07.210533: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2020-04-15 20:58:07.213099: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusolver.so.10
2020-04-15 20:58:07.214498: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusparse.so.10
2020-04-15 20:58:07.219816: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2020-04-15 20:58:07.219842: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1598] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...

**Comment by gadagashwini** [comment]
@domindominik,
As the error says Could not load dynamic library 'libcudart.so.10.1, you may have to install CUDA 10.1. 
Similar issue [#34759](https://github.com/tensorflow/tensorflow/issues/34759).

**Comment by domindominik** [comment]
@gadagashwini how to fix it without install oldest Cuda version? When Tensorflow will start supporting Cuda 10.2?

**Comment by gadagashwini** [comment]
@domindominik, To use CUDA 10.2 with Tensorflow 2.2. Please build the Tensorflow from source.
Follow the instructions mentioned [here](https://www.tensorflow.org/install/source). And also take a look at this [comment.](https://github.com/tensorflow/tensorflow/issues/38194#issuecomment-609922803)Thanks

**Comment by Saduf2019** [comment]
@domindominik,
please update on the above comment

**Comment by cyrstem** [comment]
hi  im having sort  of a different  issue but related 
` W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcudart.so.10.2'; dlerror: libcudart.so.10.2: cannot open shared object file: No such file or directory
`
my data 
 NVIDIA-SMI 430.64       Driver Version: 430.64       CUDA Version: 10.1
 if   i try  tf.config.list_physical_devices('GPU') i get :

`2020-04-30 20:33:48.029296: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcudart.so.10.2'; dlerror: libcudart.so.10.2: cannot open shared object file: No such file or directory
2020-04-30 20:33:48.029316: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2020-04-30 20:34:00.226724: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2020-04-30 20:34:00.283711: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-04-30 20:34:00.284070: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1555] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: GeForce GTX 1060 with Max-Q Design computeCapability: 6.1
coreClock: 1.48GHz coreCount: 10 deviceMemorySize: 5.94GiB deviceMemoryBandwidth: 178.99GiB/s
2020-04-30 20:34:00.284163: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcudart.so.10.2'; dlerror: libcudart.so.10.2: cannot open shared object file: No such file or directory
2020-04-30 20:34:00.285482: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2020-04-30 20:34:00.293913: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2020-04-30 20:34:00.294188: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2020-04-30 20:34:00.296217: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusolver.so.10
2020-04-30 20:34:00.297107: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusparse.so.10
2020-04-30 20:34:00.300307: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2020-04-30 20:34:00.300328: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1592] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
`

  i install  2.1.0-rc2 tensorflow via pacman   is theire a easy fix for this

**Comment by Saduf2019** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by cihilt** [comment]
https://linuxconfig.org/how-to-install-cuda-on-ubuntu-20-04-focal-fossa-linux

**Comment by legel** [comment]
See https://github.com/tensorflow/tensorflow/issues/38194
Just make sure you have installed everything (developer version) and then you can run:
`sudo apt-get install cuda-cudart-10-1`

**Comment by cyrstem** [comment]
im not in ubuntu   but  i find a solution using conda

**Comment by akarshjoice** [comment]
Using Cuda 10.1 with TensorFlow 2.2 solves the issue.

**Comment by ozett** [comment]
https://github.com/tensorflow/tensorflow/issues/38578#issuecomment-636414914  > `sudo apt-get install cuda-cudart-10-1`

install ing "cuda-cudart-10-1" seems not sufficient.. 
libcublas.so.10 still missing in my ub18. install.
i will try intstall complete "cuda-10-1" to solve this issue..

**Comment by 2008hsk** [comment]
sudo apt-get install cuda-cudart-10-1
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package cuda-cudart-10-1


It doesn't work for my case

**Comment by ozett** [comment]
try to look if you have cuda* in your repo
` apt search --names-only 'cuda'`

and lookup the packages (...10-1/10-2...) like:

![image](https://user-images.githubusercontent.com/5937811/88274208-9d77ce00-ccdb-11ea-87a5-ed951b8d53e7.png)

**Comment by ScarletMcLearn** [comment]
Same issue here.

`python
Python 3.7.8 (default, Jul 29 2020, 02:37:24) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import keras
2020-07-30 02:53:30.178267: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory
2020-07-30 02:53:30.178489: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
>>> 
`

**Comment by SkullCarverCoder** [comment]
The proper command to install cudart and avoid this error as in tensorflow 2.3.1 would be:
`sudo apt-get install libcudart10.1`

**Comment by ioannist** [comment]
Use conda for env management and just specify what cuda version you want TF to use. If you don't specify it, conda uses the latest version. The advantage of this method is that you don't have to uninstall cuda 11 or 10.2

`conda install tensorflow-gpu cudatoolkit=10.1`

**Comment by armansamimi** [comment]
> The proper command to install cudart and avoid this error as in tensorflow 2.3.1 would be:
> `sudo apt-get install libcudart10.1`

 Its work for me

**Comment by armansamimi** [comment]
> The proper command to install cudart and avoid this error as in tensorflow 2.3.1 would be:
> `sudo apt-get install libcudart10.1`

It did work for me
thanks a million.

**Comment by xxl4tomxu98** [comment]
> > The proper command to install cudart and avoid this error as in tensorflow 2.3.1 would be:
> > `sudo apt-get install libcudart10.1`
> 
> It did work for me
> thanks a million.

Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package libcudart10.1
E: Couldn't find any package by glob 'libcudart10.1'
E: Couldn't find any package by regex 'libcudart10.1'
I did exactly as above:  sudo apt-get install libcudart10.1 and got above error?

**Comment by mengna0707** [comment]
> The proper command to install cudart and avoid this error as in tensorflow 2.3.1 would be:
> `sudo apt-get install libcudart10.1`

I try your solution in my env(tensorflow=2.3.0)，and it shows the following errors：
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package libcudart10.1
E: Couldn't find any package by glob 'libcudart10.1'
E: Couldn't find any package by regex 'libcudart10.1'
have you ever seen this ？

**Comment by mengna0707** [comment]
I tried every methods given so , but still cannot  solve the problem.
Here is my environment configuration and the specific errors.

miniconda3
cuda 10.2
python  3.7.6
tensorflow 2.3.0

2020-12-13 09:55:56.519753: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2020-12-13 09:55:57.461483: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcuda.so.1
2020-12-13 09:55:57.476208: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:982] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-12-13 09:55:57.476742: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: GeForce RTX 2070 computeCapability: 7.5
coreClock: 1.62GHz coreCount: 36 deviceMemorySize: 7.79GiB deviceMemoryBandwidth: 417.29GiB/s
2020-12-13 09:55:57.476777: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2020-12-13 09:55:57.476881: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcublas.so.10'; dlerror: libcublas.so.10: cannot open shared object file: No such file or directory
2020-12-13 09:55:57.476923: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcufft.so.10'; dlerror: libcufft.so.10: cannot open shared object file: No such file or directory
2020-12-13 09:55:57.476963: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcurand.so.10'; dlerror: libcurand.so.10: cannot open shared object file: No such file or directory
2020-12-13 09:55:57.477048: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcusolver.so.10'; dlerror: libcusolver.so.10: cannot open shared object file: No such file or directory
2020-12-13 09:55:57.477102: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcusparse.so.10'; dlerror: libcusparse.so.10: cannot open shared object file: No such file or directory
2020-12-13 09:55:57.479588: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudnn.so.7
2020-12-13 09:55:57.479603: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1753] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2020-12-13 09:55:57.479778: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-12-13 09:55:57.502915: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 3192000000 Hz
2020-12-13 09:55:57.504184: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55ec536d5490 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-12-13 09:55:57.504243: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-12-13 09:55:57.507338: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1257] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-12-13 09:55:57.507386: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1263]

**Comment by MuhammadHakami** [comment]
If you have established communication with your GPU check `nvidia-smi` in bash then simply install the required cuda tools version with conda. far cleaner and can differs from environment to environment for your needs. simply execute in bash:

`conda install cudatoolkit=10.1`

**Comment by forskamse** [comment]
As per the [reference](https://tensorflow.google.cn/install/source#linux) from TensorFlow, the compiled installers of TensorFlow 2.1.0 - 2.3.0 for GPU use were based on CUDA 10.1 and cuDNN 7.6. While the most recent version 2.4.0 is starting to be compiled with CUDA 11.0 and cuDNN 8.0. 

Version | Python version | Compiler | Build tools | cuDNN | CUDA
-- | -- | -- | -- | -- | --
tensorflow-2.4.0 | 3.6-3.8 | GCC 7.3.1 | Bazel 3.1.0 | 8.0 | 11.0
tensorflow-2.3.0 | 3.5-3.8 | GCC 7.3.1 | Bazel 3.1.0 | 7.6 | 10.1
tensorflow-2.2.0 | 3.5-3.8 | GCC 7.3.1 | Bazel 2.0.0 | 7.6 | 10.1
tensorflow-2.1.0 | 2.7, 3.5-3.7 | GCC 7.3.1 | Bazel 0.27.1 | 7.6 | 10.1

So if you want to install TensorFlow 2.1.0-2.3.0 through the compiled installers (typically via pip), you must have CUDA 10.1 installed.

To check if CUDA 10.1 is correctly installed, please check:
1) run `nvidia-smi` to check whether the nvidia driver and CUDA are installed correctly at system level;
2) run `export` to see if CUDA path (usually /usr/local/cuda/) is in the environment variables;
3) run `nvcc -V` to check CUDA version.

If you have CUDA 10.2 installed instead of CUDA 10.1, the following trick may help:

```
cd /usr/local/cuda/lib64/
sudo ln -sf libcudart.so.10.2 libcudart.so.10.1
```
this creates a symbolic link from libcudart.so.10.1 to libcudart.so.10.2.

Another solution is to create an environment and install cudatoolkit 10.1 through conda, and reinstall TensorFlow in this environment. The solution provides environment level CUDA 10.1 dynamic libraries for TensorFlow.

**Comment by DKandrew** [comment]
Great found! @forskamse 
Note that your three step comments may be a bit of misleading. The `nvidia-smi` version is the driver version, as long as it is greater than 10.1 we are fine (backward compatible). The key is `nvcc -V` (the CUDA version).

**Comment by PointCloudYC** [comment]
Managed to install tensorflow-gpu 2.3 with cudatoolkit 10.1 on my cuda 10.2 driver(Jan 19th,2021)
```
conda create --name tf2 python=3.8.3
conda install cudnn==7.6.4
pip install tensorflow-gpu=2.3
```

**Comment by BryanBo-Cao** [comment]
> Use conda for env management and just specify what cuda version you want TF to use. If you don't specify it, conda uses the latest version. The advantage of this method is that you don't have to uninstall cuda 11 or 10.2
> 
> `conda install tensorflow-gpu cudatoolkit=10.1`

I don't know why but this solved my problem! Thanks! @ioannist @MuhammadHakami

**Comment by AmirthaRajan** [comment]
> cd /usr/local/cuda/lib64/
> sudo ln -sf libcudart.so.10.2 libcudart.so.10.1

Thanks a ton command solved the issue

**Comment by pkshahid** [comment]
> The proper command to install cudart and avoid this error as in tensorflow 2.3.1 would be: `sudo apt-get install libcudart10.1`

This worked for me
